docs/architecture/adr-048-consensus-fees.md

```diff
@@ -0,0 +1,181 @@
+
+#### Multi-tier pricing system
+
+[Recent IOHK blog](https://iohk.io/en/blog/posts/2021/11/26/network-traffic-and-tiered-pricing/) described a multi-tier gas pricing system. User can specify the tier of service needed for the transaction. The higher the tier, the higher priority, and higher gas price. Within the same tier, transactions pay the same gas price. It avoids the first price auction while provides prioritized service for urgent transactions.
+
+The original proposal reserve block segments for different tiers, we can't do that without modify tendermint, but it should be good enough to just specify higher priority for higher tier transactions. 
+
+The lowest tier can use constant gas price, the higher tiers adjust gas price based on block load at different speed.
+
+##### Positive
+
+- Predictable at the lowest/default tier.
+- Can prioritize urgent transactions in higher tiers.
+- Deterministic at all tiers, no price auction at all.
+
+##### Negative
+
+- TODO
+
+## Decision
+
+After comparing the trade-offs of different designs, we think the adapted multi-tier system is most suitable for cosmos-sdk's case.
+
+We can adopt EIP-1559's approach to adjust higher tier's gas prices based on block load.
+
+### Implementation
+
+We can make each tier's gas price strategy fully configurable in protocol parameters, while providing a sensible default one.
+
+Pseudocode in python-like syntax:
+
+```python
+interface TieredTx:
+  def tier(self) -> int:
+    pass
+
+def tx_tier(tx):
+    if isinstance(tx, TieredTx):
+      return tx.tier()
+    else:
+      # default tier for custom transactions
+      return 0
+
+class TierParams:
+  'gas price strategy parameters of one tier'
+  priority: int           # priority in tendermint mempool
+  initial_gas_price: Coin
+  parent_gas_target: int
+  change_speed: Decimal   # 0 means don't adjust for block load.
+
+class Params:
+    'protocol parameters'
+    tiers: List[TierParams]
+
+class State:
+    'consensus state'
+    # total gas used in last block, None when it's the first block
+    parent_gas_used: Optional[int]
+    # gas prices of last block for all tiers
+    gas_prices: List[Coin]
+
+def begin_block():
+    'Adjust gas prices'
+    for i, tier in enumerate(Params.tiers):
+        if State.parent_gas_used is None:
+            # initialized gas price for the first block
+	          State.gas_prices[i] = tier.initial_gas_price
+        else:
+            gas_used_change_rate = (State.parent_gas_used - tier.parent_gas_target) // tier.parent_gas_target
+            State.gas_prices[i] *= 1 + gas_used_change_rate * tier.change_speed
+
+def mempoolFeeTxHandler_checkTx(ctx, tx):
+    # the minimal-gas-price configured by validator, zero in deliver_tx context
+    validator_price = ctx.MinGasPrice()
+    consensus_price = State.gas_prices[tx_tier(tx)]
+    min_price = max(validator_price, consensus_price)
+
+    if tx.gas_price() < min_price:
```

Comment:
how the user will know what gas price to use? Maybe if we use `0`then we should automatically adjust and set the `min_price`?

Reply:
we can provide a query endpoint for that?

Reply:
I tried to keep exiting tx fields and semantics in this proposal, but maybe it makes more sense to change it to `gasPriceCap`, so it'll always be charged with what's computed on-chain as long as the computed price `<= gasPriceCap`. And it doesn't need to keep the per-validator `minimal-gas-price` anymore.

Reply:
updated proposal to change `gasPrice` to `gasPriceCap`, the actual gas price used is the one computed in consensus.